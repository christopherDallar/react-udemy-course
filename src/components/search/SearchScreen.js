import React, { useMemo } from 'react'
import { useLocation, useNavigate } from 'react-router-dom'
import queryString from 'query-string'

import { useForm } from './../../hooks/useForm'
import { getHeroesByName } from './../../selectors/getHeroByName'
import { HeroCard } from './../hero/HeroCard'

const ErrorMessage = ({ errorType = '', q = '' }) => (
  <>
    {errorType === 'empty' && (
      <div className="alert alert-info">Search a heroe</div>
    )}

    {errorType === 'any' && (
      <div className="alert alert-danger">No results "{q}"</div>
    )}
  </>
)

export const SearchScreen = () => {
  const navigate = useNavigate()
  const location = useLocation()

  // If q doesn't exist in the query string object, set it to an empty string
  const { q = '' } = queryString.parse(location.search)

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

  const [{ searchText }, handleInputChange] = useForm({
    searchText: q,
  })

  const handleSearch = (e) => {
    e.preventDefault()

    navigate(`?q=${searchText}`) // Without / it will take the current path and add the query params
  }

  return (
    <>
      <h1>Searcher</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Search</h4>

          <form onSubmit={handleSearch} aria-label="form">
            <input
              type="text"
              placeholder="Search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              onChange={handleInputChange}
              value={searchText}
            />

            <button className="btn btn-outline-primary mt-1" type="submit">
              Search...
            </button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {q === '' ? (
            <ErrorMessage errorType="empty" />
          ) : (
            heroesFiltered.length === 0 && (
              <ErrorMessage errorType="any" q={q} />
            )
          )}

          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  )
}
