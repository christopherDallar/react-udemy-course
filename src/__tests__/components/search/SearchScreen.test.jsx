import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter, useNavigate } from 'react-router-dom'
import { SearchScreen } from '../../../components/search/SearchScreen'

const mockedUseNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}))

describe('Testing <SearchScreen />', () => {
  test('should to match with snapshot, with ', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchScreen />
      </MemoryRouter>,
    )

    expect(container).toMatchSnapshot()
  })

  test('should to show batman and input with queryString Value ', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>,
    )

    const input = screen.getByRole('textbox')
    expect(input.value).toBe('batman')

    const img = screen.getByRole('img')
    expect(img.src).toContain('/src/assets/heroes/dc-batman.jpg')
  })

  // test('should to show error if does not exist hero (batman123)', () => {

  //  })

  test('should call navigate to new screen', () => {
    const inputValue = 'superman'

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>,
    )

    const input = screen.getByRole('textbox')

    fireEvent.change(input, {
      target: { name: 'searchText', value: inputValue },
    })

    expect(input.value).toBe(inputValue)

    const form = screen.getByLabelText('form')
    fireEvent.submit(form)
    expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`)
  })
})
