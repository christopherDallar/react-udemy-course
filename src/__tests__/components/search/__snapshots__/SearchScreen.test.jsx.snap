// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <SearchScreen /> should to match with snapshot, with  1`] = `
<div>
  <h1>
    Searcher
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-info"
      >
        Search a heroe
      </div>
    </div>
  </div>
</div>
`;
